### PR TITLE
Implement shuffle opitmization for JOIN.

### PR DESCRIPTION
**Pull request by Alex-Cheng**
### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Improve JOIN performance by introducing shuffling, controlled by settings `shuffle_join_optimization_buckets` and `shuffle_join_optimization_max_key`.

Setting `shuffle_join_optimization_buckets`: the number of buckets for shuffling optimization.
Setting `shuffle_join_optimization_max_key`: the max join key value for shuffling optimization. The shuffling optimization assumes that join key is continuous sequence of UInt64.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by UnamedRus** [comment]
> the max join key value for shuffling optimization. The shuffling optimization assumes that join key is continuous sequence of UInt64.

I'm curious what it does mean?

**Comment by Alex-Cheng** [comment]
> > the max join key value for shuffling optimization. The shuffling optimization assumes that join key is continuous sequence of UInt64.
> 
> I'm curious what it does mean? optimization for case, when our join by incrementing UIntX key?

There is a table A and table B, each table has ID as primary key and auto-increase. So the ID column is 'continuous sequence of UInt64'. Then table A has a foreign key 'fk_b' which responds to B's ID column. The query `select * from A left join B on A.fk_b = B.ID settings join_algorithm='full_sorting_merge'` will sort the A.fk_b then A.fk_b in the sorted result is continuous.

**Comment by Alex-Cheng** [comment]
img_v2_7d9a6584-cde6-4ed1-b02a-c54929d88d4g
Above is the result pipeline.

File: src/Core/Settings.h [hunk 1]
```diff
@@ -806,6 +806,8 @@ class IColumn;
     M(Timezone, session_timezone, "", "This setting can be removed in the future due to potential caveats. It is experimental and is not suitable for production usage. The default timezone for current session or query. The server default timezone if empty.", 0) \
     M(Bool, allow_create_index_without_type, false, "Allow CREATE INDEX query without TYPE. Query will be ignored. Made for SQL compatibility tests.", 0) \
     M(Bool, create_index_ignore_unique, false, "Ignore UNIQUE keyword in CREATE UNIQUE INDEX. Made for SQL compatibility tests.", 0) \
+    M(UInt64, shuffle_join_optimization_buckets, 0, "基于shuffle的join运算优化所规定的分桶数，如果是0或者1则代表不启用shuffle优化。", 0 ) \
+    M(UInt64, shuffle_join_optimization_max_key, 100000000, "基于shuffle的join运算优化所规定的连接关键字最大值，用于均分分桶范围。", 0 ) \
```
Comment (nikitamikhaylov) [inline src/Core/Settings.h:810]: :) Please write the description in English

**Comment by UnamedRus** [comment]
>  join_algorithm='full_sorting_merge' 

So, this optimization only for full sorting join?
OK

I'm just also thinking about some heuristic, which will do some kind of optimized "hash join", but using arrays instead of hashtable, if we have  "So the ID column is 'continuous sequence of UInt64'", it should be faster & more memory efficient than current hash join

File: src/Core/Settings.h (hunk 1 above)
Comment (Alex-Cheng) [reply src/Core/Settings.h:810]: I will update the description.

**Comment by Alex-Cheng** [comment]
> > join_algorithm='full_sorting_merge'
> 
> So, this optimization only for full sorting join? OK
> 
> I'm just also thinking about some heuristic, which will do some kind of optimized "hash join", but using arrays instead of hashtable, if we have "So the ID column is 'continuous sequence of UInt64'", it should be faster & more memory efficient than current hash join

I think it is a good idea. Sorry for late reply, I work on the CH community in my spare time and recently my work load is a little bit heavy.

**Comment by Alex-Cheng** [comment]
The performance of shuffle+full_sorting_merge join is fastest in my test, 100M rows LEFT JOINING 100M rows costs only 2.9 second, comparing to 'full sorting merge join'(6 second) and 'hash join'(18 second) and 'grace hash join'(20 second). 

The query is:
```SQL
SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'full_sorting_merge', max_rows_in_set_to_optimize_join = 0, shuffle_join_optimization_buckets=4,shuffle_join_optimization_max_key=100000000 format Null;
```

Performance comparing other join algorithms:

    Fully sorting merge join with in-order optimization            | 6.126s
    Grace hash join                                                                     | 20.020s
    Hash join                                                                              | 17.367s
    Auto                                                                                      | 20.799s
    Parallel hash join                                                                   | 11.705s
    Partial merge join                                                                  | 19.717s
    Fully sorting merge join with in-order and shuffle optimization     | 2.969s

Below is the details of my manual test for 'fully_sorting_merge JOIN by shuffling'. The tables for test are test_join.left_100_million and test_join.right_100_million.

```
localhost.localdomain :) show create test_join.left_100_million

SHOW CREATE TABLE test_join.left_100_million

Query id: 6d7278aa-dae1-4ae4-9f41-f9e7a1fd450b

┌─statement──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE test_join.left_100_million
(
    `uint64` UInt64,
    `string` String
)
ENGINE = MergeTree
ORDER BY uint64
SETTINGS index_granularity = 8192 │
└────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.005 sec. Peak memory: 9.04 KiB.

localhost.localdomain :) show create test_join.right_100_million

SHOW CREATE TABLE test_join.right_100_million

Query id: 884dff85-a84a-4fcb-8445-62c2c5e986d8

┌─statement───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ CREATE TABLE test_join.right_100_million
(
    `uint64` UInt64,
    `string` String
)
ENGINE = MergeTree
ORDER BY uint64
SETTINGS index_granularity = 8192 │
└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

1 row in set. Elapsed: 0.002 sec. Peak memory: 9.04 KiB.

localhost.localdomain :) select count() from  test_join.left_100_million

SELECT count()
FROM test_join.left_100_million

Query id: c601c9e7-533a-455f-81e9-28b1cf722622

┌───count()─┐
│ 100000000 │
└───────────┘

1 row in set. Elapsed: 0.002 sec. Peak memory: 29.07 KiB.

localhost.localdomain :) select count() from  test_join.right_100_million

SELECT count()
FROM test_join.right_100_million

Query id: 63d47cd0-c2aa-4d77-a796-7911d4ce2e01

┌───count()─┐
│ 100000000 │
└───────────┘

1 row in set. Elapsed: 0.002 sec. Peak memory: 29.07 KiB.

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'full_sorting_merge', max_rows_in_set_to_optimize_join = 0 format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'full_sorting_merge', max_rows_in_set_to_optimize_join = 0
FORMAT `Null`

Query id: 4d4cfff7-6904-4528-b751-b9e3feeb73c2

Ok.

0 rows in set. **Elapsed: 6.126 sec.** Peak memory: 102.05 MiB. Processed 200.00 million rows, 5.40 GB (32.65 million rows/s., 881.50 MB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'full_sorting_merge', max_rows_in_set_to_optimize_join = 0, shuffle_join_optimization_buckets=4,shuffle_join_optimization_max_key=100000000 format Null
[Alex] The above is 'JOIN by shuffling' that I implemented and the settings `shuffle_join_optimization_buckets` and `shuffle_join_optimization_max_key` are required here.

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'full_sorting_merge', max_rows_in_set_to_optimize_join = 0, shuffle_join_optimization_buckets = 4, shuffle_join_optimization_max_key = 100000000
FORMAT `Null`

Query id: f71842f6-cb1d-449c-af41-3c3e1b167bc0

Ok.

0 rows in set. **Elapsed: 2.969 sec.** Peak memory: 9.03 GiB. Processed 200.00 million rows, 5.40 GB (67.36 million rows/s., 1.82 GB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'grace_hash'  format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'grace_hash'
FORMAT `Null`

Query id: b5fb250d-6019-4baf-8918-b9260c3c8429

Ok.

0 rows in set. **Elapsed: 20.020 sec.** Peak memory: 12.43 GiB. Processed 200.00 million rows, 5.40 GB (9.99 million rows/s., 269.73 MB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'hash'  format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'hash'
FORMAT `Null`

Query id: e29b6d9f-d944-49bd-a16d-3cee909d3da9

Ok.

0 rows in set. **Elapsed: 17.367 sec.** Peak memory: 11.04 GiB. Processed 200.00 million rows, 5.40 GB (11.52 million rows/s., 310.93 MB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'auto'  format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'auto'
FORMAT `Null`

Query id: 9b7f8e1e-c1a5-426a-bdb0-35c327ab5c27

Ok.

0 rows in set. **Elapsed: 20.799 sec.** Peak memory: 1.71 GiB. Processed 200.00 million rows, 5.40 GB (9.62 million rows/s., 259.63 MB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'parallel_hash'  format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'parallel_hash'
FORMAT `Null`

Query id: 710e5def-9dbc-4f85-9c00-279b60c57041

Ok.

0 rows in set. **Elapsed: 11.705 sec.** Peak memory: 11.05 GiB. Processed 200.00 million rows, 5.40 GB (17.09 million rows/s., 461.34 MB/s.)

localhost.localdomain :) SELECT * FROM test_join.left_100_million AS l
                             LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
                             SETTINGS join_algorithm = 'partial_merge'  format Null

SELECT *
FROM test_join.left_100_million AS l
LEFT JOIN test_join.right_100_million AS r ON l.uint64 = r.uint64
SETTINGS join_algorithm = 'partial_merge'
FORMAT `Null`

Query id: d01e1261-ef0b-419e-9ddf-a5b31e1c932f

Ok.

0 rows in set. **Elapsed: 19.717 sec.** Peak memory: 1.71 GiB. Processed 200.00 million rows, 5.40 GB (10.14 million rows/s., 273.88 MB/s.)
```

The problem is the implementation of JOIN with shuffling only fits for a few of specific case where join key is a single UInt64 column and the values of the column is continuous and ordered, because continuous and ordered is good for shuffing(i.e. in most cases, it shuffle a whole block into a single bucket a time). The reason that it requires UInt64 column is that the shuffling algorithm is simply segmentation(e.g. 0~10000->bucket1, 10000~20000 -> bucket2 ...).

I am thinking about how to broaden the applicable scope of the optimization. Therefore it is still a draft PR.

**Comment by UnamedRus** [comment]
> The problem is the implementation of JOIN with shuffling only fits for a few of specific case where join key is a single UInt64 

But, it's somewhat common case, because people are coming with their data from other DBMS with auto increment columns. 
Plus, some benchmark also generate data in that way.

**Comment by alexey-milovidov** [comment]
Why closing?

**Comment by alexey-milovidov** [comment]
I see https://github.com/ClickHouse/ClickHouse/pull/70777

**Pull request closed without merging**